rest-server/templates/src/main/scala/com/actionml/templates/cb/CBAlgorithm.scala

```diff
@@ -32,60 +39,87 @@ import scala.concurrent.Future
   * of events when a new one is detected, then updating the model for that group for subsequent queries.
   * The GroupTrain Actors are managed by the CBAlgorithm and will be added and killed when needed.
   *
-  * @param p All needed params for the CB lib as well as the dataset containing events and data used in training.
-  *          The dataset will contain groups from which the GroupTrain Actors are created
   */
-class CBAlgorithm(p: CBAlgoParams) extends Algorithm(new Mongo, p: CBAlgoParams) {
-
-  private val system = ActorSystem("CBAlgorithm")
+class CBAlgorithm(dataset: CBDataset) extends Algorithm(new Mongo) with JsonParser {
 
+  private val actors = ActorSystem("CBAlgorithm")
   private var trainers = Map.empty[String, ActorRef]
+  var params: CBAlgoParams = _
 
   // from the Dataset determine which groups are defined and start training on them
 
-  def init(): CBAlgorithm = {
-    val groups: Map[String, MongoCollection] = p.dataset.CBCollections.usageEventGroups
-    logger.trace(s"Init manager for ${groups.size} groups. ${groups.mkString(", ")}")
-    val exists = trainers.keys.toList
-    val diff = groups.filterNot { case (key, _) ⇒
-      exists.contains(key)
-    }
-
-    diff.foreach { case (trainer, collection) ⇒
-      val actor = system.actorOf(SingleGroupTrainer.props(collection), trainer)
-      trainers += trainer → actor
-    }
-    this
+  def init(json: String): Validated[ValidateError, Boolean] = {
+    val response = parseAndValidate[CBAlgoParams](json)
+
+    if (response.isValid) {
+      params = response.getOrElse(CBAlgoParams())
+
+      val groups: Map[String, MongoCollection] = dataset.usageEventGroups
+      logger.trace(s"Init manager for ${groups.size} groups. ${groups.mkString(", ")}")
+      val exists = trainers.keys.toList
+      val diff = groups.filterNot { case (key, _) =>
+        exists.contains(key)
+      }
+
+      diff.foreach { case (trainer, collection) =>
+        val actor = actors.actorOf(SingleGroupTrainer.props(collection), trainer)
+        trainers += trainer → actor
+      }
+      Valid(true)
+    } else response.map(_ => false)
   }
 
   def train(groupName: String): Unit = {
-    logger.trace("Train trainer {}", groupName)
-    trainers(groupName) ! SingleGroupTrainer.Train
+    try {
+      logger.trace("Train trainer {}", groupName)
+      trainers(groupName) ! SingleGroupTrainer.Train
+    } catch{
+      case e: NoSuchElementException =>
+        logger.error(s"Training triggered on non-existent group: $groupName The group must be initialized first. " +
+          s"All events for this group will be ignored. ")
+    }
   }
```

Comment:
using exceptions for existence check is bad form IMO, check the keyset of the Map instead.